File: java-impl/src/main/java/hackathonlib/IGameManager.java
package hackathonlib;

import java.util.List;

/**
 * This interface must be implemented by servers and clients.
 * It specifies how to react on different packets.
 * Some are only relevant for clients some only for servers.
 *
 * @author Michael Fürst
 * @version 1.0
 * @since 23.11.2014
 */
public interface IGameManager {
    /**
     * Notifies a server when a connection is started. (Clients can ignore this.)
     * @param connection The connection.
     */
    void connectionStarted(Connection connection);

    /**
     * Notifies a server when a connection is ready. (Clients can ignore this.)
     * @param connection The connection.
     */
    void connectionReady(Connection connection);

    /**
     * Notifies a server when a connection is lost. (Clients can ignore this.)
     * @param connection The connection.
     */
    void connectionLost(Connection connection);

    /**
     * Notifies a server when a connection wants to execute a command. (Clients can ignore this.)
     * @param connection The connection.
     * @param p The command.
     */
    void executeCommand(Connection connection, String p);

    /**
     * Notifies a server when a connection wants to change it's speed. (Clients can ignore this.)
     * @param connection The connection.
     * @param p The speed.
     */
    void changeSpeed(Connection connection, double p);

    /**
     * Notifies a server when a connection wants to change it's rotation. (Clients can ignore this.)
     * @param connection The connection.
     * @param p The rotation.
     */
    void changeRotation(Connection connection, double p);

    /**
     * Update the scene. (For clients. Servers should just ignore this.)
     * @param scene The scene.
     */
    void updateScene(List<GameObject> scene);

    /**
     * The scene. (Only for servers.)
     * @return The scene. (Clients should return null.)
     */
    List<GameObject> getScene();

    /**
     * The command to execute.
     * @return The current command.
     */
    String getCommand();

    /**
     * Your name.
     * @return Your name.
     */
    String getName();

    /**
     * Return null if you do not want to set a speed, otherwise return the speed.
     * @return The speed or null.
     */
    Double getSpeed();

    /**
     * Return null if you do not want to set a rotation, otherwise return the rotation.
     * @return The rotation or null.
     */
    Double getRotation();

    /**
     * Do you want to ping the server this frame.
     * Set this true if all other outputs are null, otherwise you will loose your connection.
     * @return True for a ping.
     */
    boolean doPing();
}

File: csharp-impl/HackathonLib/IGameManager.cs
using System.Collections.Generic;

namespace HackathonLib
{
    public interface IGameManager
    {
        /// <summary>
        /// Called when a connection is successfully established.
        /// (Recommended for servers)
        /// </summary>
        /// <param name="connection"></param>
        void ConnectionStarted(Connection connection);

        /// <summary>
        /// Called on servers when a connection is ready to join a game. Means that the connection has a valid playername.
        /// (Recommended for servers)
        /// </summary>
        /// <param name="connection"></param>
        void ConnectionReady(Connection connection);

        /// <summary>
        /// Called when the connection is lost.
        /// (Recommended for servers)
        /// </summary>
        /// <param name="connection"></param>
        void ConnectionLost(Connection connection);

        /// <summary>
        /// Called when the other side of the network wishes your implementation to execute the given command.
        /// (Recommended for servers)
        /// </summary>
        /// <param name="connection"></param>
        /// <param name="command">The command that should be executed.</param>
        void ExecuteCommand(Connection connection, string command);

        /// <summary>
        /// Called when the other side of the network wishes to change the movement speed.
        /// (Recommended for servers)
        /// </summary>
        /// <param name="connection"></param>
        /// <param name="speed"></param>
        void ChangeSpeed(Connection connection, double speed);

        /// <summary>
        /// Called when the other side of the network wishes to change the rotation.
        /// (Recommended for servers)
        /// </summary>
        /// <param name="connection"></param>
        /// <param name="rotation"></param>
        void ChangeRotation(Connection connection, double rotation);

        /// <summary>
        /// This contains the current scene.
        /// (Recommended for servers)
        /// </summary>
        /// <returns>The current scene or null.</returns>
        List<GameObject> GetScene();

        /// <summary>
        /// Update the scene to fit to the network.
        /// </summary>
        /// <param name="scene">The current scene or null.</param>
        void UpdateScene(List<GameObject> scene);

        /// <summary>
        /// Return a command if one should be transmitted.
        /// </summary>
        /// <returns>The command or null.</returns>
        string GetCommand();

        /// <summary>
        /// Return the name of the player/connection.
        /// </summary>
        /// <returns>The name or null.</returns>
        string GetName();

        /// <summary>
        /// Return the speed if one should be transmitted.
        /// </summary>
        /// <returns>The speed or null.</returns>
        double? GetSpeed();

        /// <summary>
        /// Return the rotation if one should be transmitted.
        /// </summary>
        /// <returns>The rotation or null.</returns>
        double? GetRotation();

        /// <summary>
        /// Return if the ping flag should be set.
        /// </summary>
        /// <returns>If a ping should be executed.</returns>
        bool DoPing();
    }
}
